Build error with Postgres because of libxml2

Since a few days, Postgres build fails because libxml2 isn't detected (while it looks present):


```
configure: error: library 'xml2' (version >= 2.6.23) is required for XML support
checking for xmlSaveToBuffer in -lxml2... no
```

https://github.com/dunglas/frankenphp/actions/runs/8631360668/job/23659621879?pr=712

In the `config.log` of Postgres, we can see that the problem looks related to ICU:

```
configure:13082: checking for xmlSaveToBuffer in -lxml2
configure:13107: clang -o conftest -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Werror=unguarded-availability-new -Wendif-labels -Wmissing-format-attribute -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk  -DLIBXML_STATIC -I/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/include/libxml2 -I/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/include  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.4.sdk  -L/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot//Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib -L/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib  conftest.c -lxml2  -lssl -lcrypto -lz -lreadline -ltermcap -lm  >&5
ld: warning: search path '/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot//Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib' not found
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[2](buf.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[4](catalog.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[5](chvalid.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[6](debugXML.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[7](dict.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[8](encoding.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[9](entities.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[10](error.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[11](globals.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[12](hash.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[13](HTMLparser.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[14](HTMLtree.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[16](list.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[18](nanohttp.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[19](parser.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[20](parserInternals.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[21](pattern.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[22](relaxng.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[24](SAX2.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[26](threads.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[27](tree.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[28](uri.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[29](valid.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[30](xinclude.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[32](xmlIO.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[33](xmlmemory.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[35](xmlreader.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[36](xmlregexp.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[37](xmlsave.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[38](xmlschemas.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[39](xmlschemastypes.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[40](xmlstring.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[41](xmlunicode.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[43](xpath.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
ld: warning: object file (/Users/dunglas/workspace/frankenphp/dist/static-php-cli/buildroot/lib/libxml2.a[44](xpointer.c.o)) was built for newer 'macOS' version (14.3) than being linked (14.0)
Undefined symbols for architecture arm64:
  "_UCNV_FROM_U_CALLBACK_STOP_74", referenced from:
      _openIcuConverter in libxml2.a[8](encoding.c.o)
  "_UCNV_TO_U_CALLBACK_STOP_74", referenced from:
      _openIcuConverter in libxml2.a[8](encoding.c.o)
  "_libiconv", referenced from:
      _xmlIconvWrapper in libxml2.a[8](encoding.c.o)
  "_libiconv_close", referenced from:
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlCharEncCloseFunc in libxml2.a[8](encoding.c.o)
      _xmlCharEncCloseFunc in libxml2.a[8](encoding.c.o)
      ...
  "_libiconv_open", referenced from:
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
      _xmlFindCharEncodingHandler in libxml2.a[8](encoding.c.o)
  "_ucnv_close_74", referenced from:
      _openIcuConverter in libxml2.a[8](encoding.c.o)
      _closeIcuConverter in libxml2.a[8](encoding.c.o)
      _closeIcuConverter in libxml2.a[8](encoding.c.o)
      _xmlCharEncCloseFunc in libxml2.a[8](encoding.c.o)
      _xmlCharEncCloseFunc in libxml2.a[8](encoding.c.o)
      _xmlCharEncCloseFunc in libxml2.a[8](encoding.c.o)
      _xmlCharEncCloseFunc in libxml2.a[8](encoding.c.o)
      ...
  "_ucnv_convertEx_74", referenced from:
      _xmlEncInputChunk in libxml2.a[8](encoding.c.o)
      _xmlEncOutputChunk in libxml2.a[8](encoding.c.o)
  "_ucnv_open_74", referenced from:
      _openIcuConverter in libxml2.a[8](encoding.c.o)
      _openIcuConverter in libxml2.a[8](encoding.c.o)
  "_ucnv_setFromUCallBack_74", referenced from:
      _openIcuConverter in libxml2.a[8](encoding.c.o)
  "_ucnv_setToUCallBack_74", referenced from:
      _openIcuConverter in libxml2.a[8](encoding.c.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Note: the problem seems to also happen on Linux.

--- a/src/SPC/builder/linux/library/openssl.php
+++ b/src/SPC/builder/linux/library/openssl.php
@@ -24,6 +24,7 @@ use SPC\builder\linux\SystemUtil;
 use SPC\exception\FileSystemException;
 use SPC\exception\RuntimeException;
 use SPC\exception\WrongUsageException;
+use SPC\store\FileSystem;
 
 class openssl extends LinuxLibraryBase
 {
@@ -76,5 +77,15 @@ class openssl extends LinuxLibraryBase
             ->exec("make -j{$this->builder->concurrency} CNF_EX_LIBS=\"{$ex_lib}\"")
             ->exec("make install_sw DESTDIR={$destdir}");
         $this->patchPkgconfPrefix(['libssl.pc', 'openssl.pc', 'libcrypto.pc']);
+        // patch for openssl 3.3.0+
+        if (!str_contains($file = FileSystem::readFile(BUILD_LIB_PATH . '/pkgconfig/libssl.pc'), 'prefix=')) {
+            FileSystem::writeFile(BUILD_LIB_PATH . '/pkgconfig/libssl.pc', 'prefix=' . BUILD_ROOT_PATH . "\n" . $file);
+        }
+        if (!str_contains($file = FileSystem::readFile(BUILD_LIB_PATH . '/pkgconfig/openssl.pc'), 'prefix=')) {
+            FileSystem::writeFile(BUILD_LIB_PATH . '/pkgconfig/openssl.pc', 'prefix=' . BUILD_ROOT_PATH . "\n" . $file);
+        }
+        if (!str_contains($file = FileSystem::readFile(BUILD_LIB_PATH . '/pkgconfig/libcrypto.pc'), 'prefix=')) {
+            FileSystem::writeFile(BUILD_LIB_PATH . '/pkgconfig/libcrypto.pc', 'prefix=' . BUILD_ROOT_PATH . "\n" . $file);
+        }
     }
 }

--- a/src/SPC/builder/macos/library/openssl.php
+++ b/src/SPC/builder/macos/library/openssl.php
@@ -23,6 +23,7 @@ namespace SPC\builder\macos\library;
 use SPC\exception\FileSystemException;
 use SPC\exception\RuntimeException;
 use SPC\exception\WrongUsageException;
+use SPC\store\FileSystem;
 
 class openssl extends MacOSLibraryBase
 {
@@ -58,5 +59,15 @@ class openssl extends MacOSLibraryBase
             ->exec("make -j{$this->builder->concurrency} CNF_EX_LIBS=\"{$ex_lib}\"")
             ->exec("make install_sw DESTDIR={$destdir}");
         $this->patchPkgconfPrefix(['libssl.pc', 'openssl.pc', 'libcrypto.pc']);
+        // patch for openssl 3.3.0+
+        if (!str_contains($file = FileSystem::readFile(BUILD_LIB_PATH . '/pkgconfig/libssl.pc'), 'prefix=')) {
+            FileSystem::writeFile(BUILD_LIB_PATH . '/pkgconfig/libssl.pc', 'prefix=' . BUILD_ROOT_PATH . "\n" . $file);
+        }
+        if (!str_contains($file = FileSystem::readFile(BUILD_LIB_PATH . '/pkgconfig/openssl.pc'), 'prefix=')) {
+            FileSystem::writeFile(BUILD_LIB_PATH . '/pkgconfig/openssl.pc', 'prefix=' . BUILD_ROOT_PATH . "\n" . $file);
+        }
+        if (!str_contains($file = FileSystem::readFile(BUILD_LIB_PATH . '/pkgconfig/libcrypto.pc'), 'prefix=')) {
+            FileSystem::writeFile(BUILD_LIB_PATH . '/pkgconfig/libcrypto.pc', 'prefix=' . BUILD_ROOT_PATH . "\n" . $file);
+        }
     }
 }

--- a/src/SPC/builder/unix/library/postgresql.php
+++ b/src/SPC/builder/unix/library/postgresql.php
@@ -27,26 +27,31 @@ trait postgresql
             'libxslt' => 'libxslt',
             'icu' => 'icu-i18n',
         ];
+        $error_exec_cnt = 0;
 
         foreach ($optional_packages as $lib => $pkg) {
             if ($this->getBuilder()->getLib($lib)) {
                 $packages .= ' ' . $pkg;
                 $output = shell()->execWithResult("pkg-config --static {$pkg}");
+                $error_exec_cnt += $output[0] === 0 ? 0 : 1;
                 logger()->info(var_export($output[1], true));
             }
         }
 
         $output = shell()->execWithResult("pkg-config --cflags-only-I --static {$packages}");
+        $error_exec_cnt += $output[0] === 0 ? 0 : 1;
         if (!empty($output[1][0])) {
             $cppflags = $output[1][0];
             $envs .= " CPPFLAGS=\"{$cppflags}\"";
         }
         $output = shell()->execWithResult("pkg-config --libs-only-L --static {$packages}");
+        $error_exec_cnt += $output[0] === 0 ? 0 : 1;
         if (!empty($output[1][0])) {
             $ldflags = $output[1][0];
             $envs .= $this instanceof MacOSLibraryBase ? " LDFLAGS=\"{$ldflags}\" " : " LDFLAGS=\"{$ldflags} -static\" ";
         }
         $output = shell()->execWithResult("pkg-config --libs-only-l --static {$packages}");
+        $error_exec_cnt += $output[0] === 0 ? 0 : 1;
         if (!empty($output[1][0])) {
             $libs = $output[1][0];
             $libcpp = '';
@@ -54,6 +59,9 @@ trait postgresql
                 $libcpp = $this instanceof LinuxLibraryBase ? ' -lstdc++' : ' -lc++';
             }
             $envs .= " LIBS=\"{$libs}{$libcpp}\" ";
+        }
+        if ($error_exec_cnt > 0) {
+            throw new RuntimeException('Failed to get pkg-config information!');
         }
 
         FileSystem::resetDir($this->source_dir . '/build');

--- a/src/globals/test-extensions.php
+++ b/src/globals/test-extensions.php
@@ -13,7 +13,7 @@ declare(strict_types=1);
 
 // If you want to test your added extensions and libs, add below (comma separated, example `bcmath,openssl`).
 $extensions = match (PHP_OS_FAMILY) {
-    'Linux', 'Darwin' => 'ds,simdjson',
+    'Linux', 'Darwin' => 'pgsql,intl,xml,openssl',
     'Windows' => 'mbstring,pdo_sqlite,mbregex,ffi,ds,simdjson',
 };
 
@@ -27,7 +27,7 @@ $with_libs = match (PHP_OS_FAMILY) {
 // You can use `common`, `bulk`, `minimal` or `none`.
 // note: combination is only available for *nix platform. Windows must use `none` combination
 $base_combination = match (PHP_OS_FAMILY) {
-    'Linux', 'Darwin' => 'common',
+    'Linux', 'Darwin' => 'minimal',
     'Windows' => 'none',
 };
 

